Add Service Claim documentation (#51)
primaza/primaza: docs/entities/serviceclaim.md

# ServiceClaim

> A Service Claim represents a claim for service.

Reviewer: for a service?

## Specification

The definition of a ServiceClaim can be obtained directly from our [ServiceClaim
CRD](../../config/crd/bases/primaza.io_serviceclaims.yaml). The specification
contains two required properties:

- ServiceClassIdentity: A set of key/value pairs that identify the service
  class. Examples of service class identity keys include type of service, and
  provider of service. This property is required.
- ServiceEndpointDefinitionKeys: An array of keys that is required for
  connectivity. The values corresponding to each of these keys will be extracted
  from the service. This property is required.

There are other three optional fields. The EnvironmentTag and
ApplicationClusterContext are mutually exclusive:

- Application: Fields indentifies application resources through kind, apiVersion and label selector. 
- EnvironmentTag: A string representing one of the environment.

> - ApplicationClusterContext:

Reviewer: Define this

Author: Done

## Status

The Status of the ServiceClaim is also defined under the [ServiceClaim

> CRD](../../config/crd/bases/primaza.io_serviceclaims.yaml). It contains only one

Reviewer: I thought you have more properties to set the registered service resource the matched

Author: This is the only `required` field. I have mentioned others also.

required property to track state. The state could be either `Pending` or
`Resolved`. If the state is `Resolved`, there should be Secret and Service
Binding resources created

There is a `claimID` fields with a unique ID for the claim.  And there is another field, `registeredService` that points the service.

## Use Cases

### Creation

> When a Service Claim is created, Primaza should find a service based on the given input and create Secret and Service Binding resources.  Then it should the the state accordingly.

Reviewer: I think you should expand this on what find a service means. Not to much details but the fact that we are going to try to match SCI and SED Keys with a Registered service.

### Deletion

> Currently no operations is performed.

Reviewer: Forget about currently, write what it should be, this will help you understand what need to be implemented

Author: Done

### Update

> When a Service Claim is updated, Primaza should find a service based on the given input and create Secret and Service Binding resources.  Then it should the the state accordingly.

Reviewer 2: Also, eventually we should detail here which field can be updated or not

```suggestion
When a Service Claim is updated, Primaza should update the Service Endpoint Definition Secret, the Service Binding and the Service Claim's state accordingly.
```

# ServiceClaim

> A Service Claim represents a claim for Registered Service.

Reviewer: for _a_ Registered Service?

Author: Done

## Specification

The definition of a ServiceClaim can be obtained directly from our [ServiceClaim
CRD](../../config/crd/bases/primaza.io_serviceclaims.yaml). The specification
contains two required properties:

- ServiceClassIdentity: A set of key/value pairs that identify the service
  class. Examples of service class identity keys include type of service, and
  provider of service. This property is required.
- ServiceEndpointDefinitionKeys: An array of keys that is required for
  connectivity. The values corresponding to each of these keys will be extracted
  from the service. This property is required.

There are other three optional fields. The EnvironmentTag and
ApplicationClusterContext are mutually exclusive:

- Application: Fields indentifies application resources through kind, apiVersion and label selector. 
- EnvironmentTag: A string representing one of the environment.

> - ApplicationClusterContext: A combination of environment name and namespaces.

Reviewer: Hum... not sure that is what it is. It is specific cluster and namespace. It is mutually exclusive with Environment Tag that points to multiple clusters and namespaces.

Author: How about `A combination of ClusterEnvironment resource name and namespace`?

## Status

The Status of the ServiceClaim is also defined under the [ServiceClaim
CRD](../../config/crd/bases/primaza.io_serviceclaims.yaml). It contains only one
required property to track the state. The state could be either `Pending` or
`Resolved`. If the state is `Resolved`, there should be Secret and Service
Binding resources created

There is an optional `claimID` field with a unique ID for the claim.  And there is

> another optional field, `registeredService` that points the service.

Reviewer: Claim ID can be optional, but I don't thing registeredService should be optional. How would you implemented finalizer that changes state of registered service when a claim is deleted?

Author: I made that field mandatory in this PR itself.

## Use Cases

### Creation

> When a Service Claim is created, Primaza should find a Registered Service based on Service Class Identity and Service Endpoint Definition Keys and create Secret and Service Binding resources. Then it should update the state accordingly.

Reviewer: Use "will" instead of "should". Also, it will update both the state of the ServiceClaim resource and the state of the Registered Service

Author: Done

### Deletion

> When a Service Claim is deleted, Primaza should delete the Service Endpoint Definition Secret and the Service Binding.

Reviewer: It should also change the state of the Registered Service

Author: Done

### Update

> When a Service Claim is updated, Primaza should update the Service Endpoint Definition Secret, the Service Binding and the Service Claim's state accordingly.

Reviewer: I think is missing the scenario when service claim changes make the claim unmatchable. Think about changing SCI or SED in a way that does not yield a RegisteredService match.

Author: I have updated the creation and update with this match not found scenario.
